### ISSUE TITLE
find out the root cause of failing to separate ridges and transforms

### ISSUE THREAD

**Issue by michaelchin** [GPlates/gplately#228]
some MOR lines cannot be decided if it is a ridge or transform because we cannot generate a stage rotation for various reasons, such as 

- not have left/right plate IDs or reconstruction/conjugate plate IDs
 - have spreading time  in distant past or future
 - have zero stage rotation from 'spreading_time + 1' to 'spreading_time'

Need to investigate and find out the root causes.

**Comment by jcannon-gplates** [comment]
Yes I got the impression Dietmar wants a reason why each MOR feature failed - like you enumerated - so users can go in and fix the model (eg, add left/right plate IDs here; fix rotation file to avoid zero stage rotation there).

Probably the easiest (but a bit hacky I suppose) is to insert a debug statement where ever there's a `return` statement in `separate_ridge_transform_segments.get_stage_rotation_for_reconstructed_geometry()`. Then you can plot a model over time and collect the debug output and give it to someone to fix the model.

So the places you could insert are:

- [here](https://github.com/GPlates/gplately/blob/a29e3aee09271b981a834ab95985f8a2a16baece/gplately/ptt/separate_ridge_transform_segments.py#L356) (probably never happens)
  ```
  print('Time is distant past/future')
  ```
- [here](https://github.com/GPlates/gplately/blob/a29e3aee09271b981a834ab95985f8a2a16baece/gplately/ptt/separate_ridge_transform_segments.py#L366)
  ```
  print('Feature {} has no reconstruction/conjugate plate pair'.format(spreading_feature.get_name()))
  ```
- [here](https://github.com/GPlates/gplately/blob/a29e3aee09271b981a834ab95985f8a2a16baece/gplately/ptt/separate_ridge_transform_segments.py#L427)
  ```
  print('Feature {} has zero stage rotation for reconstruction/conjugate plate pair {} {} at spreading time {}'.format(
    spreading_feature.get_name(),
    reconstruction_plate_id,
    conjugate_plate_id,
    spreading_time))
  ```
- [here](https://github.com/GPlates/gplately/blob/a29e3aee09271b981a834ab95985f8a2a16baece/gplately/ptt/separate_ridge_transform_segments.py#L443)
  ```
  print('Feature {} has no left/right plate pair'.format(spreading_feature.get_name()))
  ```
- [here](https://github.com/GPlates/gplately/blob/a29e3aee09271b981a834ab95985f8a2a16baece/gplately/ptt/separate_ridge_transform_segments.py#L505)
  ```
  print('Feature {} has zero stage rotation for left/right plate pair {} {} at spreading time {}'.format(
    spreading_feature.get_name(),
    left_plate_id,
    right_plate_id,
    spreading_time))
  ```

And perhaps to help users better locate the feature you could perhaps also print (in addition to the feature name) its feature ID (`spreading_feature.get_feature_id().get_string()`) and maybe its time period (`spreading_feature.get_valid_time()`).

Having it as developer debug code has its downsides, such as users can't turn it on/off, but it saves you having to propagate it as a user option up into the user-level plotting function.

**Comment by michaelchin** [comment]
> Yes I got the impression Dietmar wants a reason why each MOR feature failed - like you enumerated - so users can go in and fix the model (eg, add left/right plate IDs here; fix rotation file to avoid zero stage rotation there).
> 
> Probably the easiest (but a bit hacky I suppose) is to insert a debug statement where ever there's a `return` statement in `separate_ridge_transform_segments.get_stage_rotation_for_reconstructed_geometry()`. Then you can plot a model over time and collect the debug output and give it to someone to fix the model.
> 
> So the places you could insert are:
> 
> * [here](https://github.com/GPlates/gplately/blob/a29e3aee09271b981a834ab95985f8a2a16baece/gplately/ptt/separate_ridge_transform_segments.py#L356) (probably never happens)
>   ```
>   print('Time is distant past/future')
>   ```
> * [here](https://github.com/GPlates/gplately/blob/a29e3aee09271b981a834ab95985f8a2a16baece/gplately/ptt/separate_ridge_transform_segments.py#L366)
>   ```
>   print('Feature {} has no reconstruction/conjugate plate pair'.format(spreading_feature.get_name()))
>   ```
> * [here](https://github.com/GPlates/gplately/blob/a29e3aee09271b981a834ab95985f8a2a16baece/gplately/ptt/separate_ridge_transform_segments.py#L427)
>   ```
>   print('Feature {} has zero stage rotation for reconstruction/conjugate plate pair {} {} at spreading time {}'.format(
>     spreading_feature.get_name(),
>     reconstruction_plate_id,
>     conjugate_plate_id,
>     spreading_time))
>   ```
> * [here](https://github.com/GPlates/gplately/blob/a29e3aee09271b981a834ab95985f8a2a16baece/gplately/ptt/separate_ridge_transform_segments.py#L443)
>   ```
>   print('Feature {} has no left/right plate pair'.format(spreading_feature.get_name()))
>   ```
> * [here](https://github.com/GPlates/gplately/blob/a29e3aee09271b981a834ab95985f8a2a16baece/gplately/ptt/separate_ridge_transform_segments.py#L505)
>   ```
>   print('Feature {} has zero stage rotation for left/right plate pair {} {} at spreading time {}'.format(
>     spreading_feature.get_name(),
>     left_plate_id,
>     right_plate_id,
>     spreading_time))
>   ```
> 
> And perhaps to help users better locate the feature you could perhaps also print (in addition to the feature name) its feature ID (`spreading_feature.get_feature_id().get_string()`) and maybe its time period (`spreading_feature.get_valid_time()`).
> 
> Having it as developer debug code has its downsides, such as users can't turn it on/off, but it saves you having to propagate it as a user option up into the user-level plotting function.

There is very helpful. Thanks John.

**Comment by jcannon-gplates** [comment]
I tried introducing a ridge-like value (in range [0,1]) for ridge lengths in d6b3704:

> Instead of separating ridge and transform segments of mid-ocean ridges. This is actually just obliquity (of spreading directions vs segment direction). A better solution might be to look at divergence at all plate boundary types (not just mid-ocean ridges).

...in the hope that crustal production/destruction rates would be balanced (in the Muller 2022 model). But it actually made it worse. So the original idea of looking at all plate boundaries (not just mid-ocean ridges and subduction zones) and adding divergence/convergence to pyGPlates first, and then using it in GPlately, is the next step. In the meantime Dietmar requested we set the ridge-transform separation angle to 70 degrees (after the testing done in #220).

![image](https://github.com/user-attachments/assets/24fef405-dee8-4d54-aa7e-744551ef5774)

**Comment by jcannon-gplates** [comment]
The crustal production/destruction rate can now be obtained by the new function `PlateReconstruction.crustal_production_destruction_rate()` added in #316. The results are much better...

![Image](https://github.com/user-attachments/assets/f9033e27-2ad8-4a5d-960b-a40ebc373c01)

This is better than using `PlateReconstruction.tessellate_mid_ocean_ridges()` and `PlateReconstruction.tessellate_subduction_zones()` (which just look at mid-ocean ridges and subduction zones).

For example, `PlateReconstruction.tessellate_subduction_zones()` requires a subduction polarity (to determine overriding plate) but there might be convergent plate boundaries that are *not* subduction zones (and so crustal destruction was only getting counted at subduction zones).

Similarly situation for `PlateReconstruction.tessellate_mid_ocean_ridges()`. But here, the problem was that spreading rates required left/right plate IDs (eg, required a mid-ocean ridge feature). But there might be divergent plate boundaries that are *not* mid-ocean ridges (and so crustal production was only getting counted at mid-ocean ridges). Also there was the issue of ridge-transform separation (ie, the 70 degree angle) - that's avoided by looking purely at plate divergence/convergence (in the new function).

We can still use `PlateReconstruction.tessellate_mid_ocean_ridges()` with a default ridge-transform separation angle of 70 degrees for other things, like spreading rates at mid-ocean ridges and mid-ocean ridge lengths. But the new function is better for general crustal production/destruction rates.
